core/src/sumeragi/main_loop.rs

```diff
@@ -352,11 +337,22 @@ impl<F: FaultInjection> SumeragiWithFault<F> {
                             // Omit signature verification during genesis round
                             block.commit_unchecked().into()
                         }
-                        Message::BlockSyncUpdate(BlockSyncUpdate { block }) => {
+                        Message::BlockSyncUpdate(block_sync_update) => {
                             // Omit signature verification during genesis round
-                            match block.revalidate_hashes() {
-                                Ok(block) => block,
-                                Err((_, error)) => {
+                            match block_sync_update.validate_and_extract_block(
+                                &self.transaction_validator,
+                                state.wsv.clone(),
+                                state.latest_block_hash,
+                                state.latest_block_height,
+                            ) {
+                                Ok(block) => match revalidate_hashes(block) {
```

Comment:
Hashes are already checked inside `validate_and_extract` block, so we can remove this call.

Reply:
True! And the function can be removed as well.